### ISSUE TITLE
Required fields around site, most importantly on Conf reg (ticket) form

### ISSUE THREAD

**Issue by jdearie** [Drupal4Gov/Drupal-GovCon-2017#292]
_This issue is specifically about the Conf reg (ticket) form, but most of the notes are applicable sitewide._

### Requirements
- _[MOVED to separate GH issue: #317 ]_ "Attendee email" field: Keep as "required," and update help text to be:
    > Enter the e-mail address of the attendee who will use this ticket. If you are registering for someone else, please enter their email address.
    - _See "ALSO..." below_
- [ ] Fix fishyness with "please do this field" popup -- or disable it -- or disable it and add a message at the very top of the form like, "Please fill in all required fields" so ppl understand why they're still stuck on the form / why it didn't submit -- or do this last suggestion but put the message at the bottom right above the submit button.
   - _See "Issues" > # 2 and its sub-bullets_
- [ ] Make required checkbox AND text fields that were missed (when user submits form) more conspicuous -- right now, only missed-required select lists get "red halo."
   - _See "Issues" > # 3_
   - [ ] "please check this box" popup for Participation Release checkbox field is NOT showing up in Jess's tests on Chrome and IE (Alison saw the popup in the wrong place per screenshot above, on Firefox) -- **_see "Issues" > # 2-ii-a_**
   - [ ] Maybe make overall "you missed this required field" styling more conspicuous, b/c it's pretty subtle :( (like, heavier highlighting or something?)
      - _See "Issues" > # 2-iii_
- [ ] Make "required" asterisk on all required fields more conspicuous (in form's default state / before user submits form)!

### Issues
_(from Alison)_
**NOTE:** Details are from testing in FF on Windows 7 (but it sounds like Jess had a comparable experience on Mac).

1. Required select lists behave fine when left blank, but **checkboxes AND text fields** are no bueno -- no red halo/highlighting.
2. There's only one of these popups, regardless of how many missed required fields.  Also, it disappears as soon as you scroll, so, like I thought maybe it was "pointing" at a field further up on the page, but I'll never know b/c it disappeared as soon as I tried to look haha.
_[note: updated this explanation and its sub-bullets b/c see UPDATE below]_
   1. **UPDATE:** The placement issue seems to MOSTLY be from the admin toolbar, b/c it isn't a problem when I'm just an authenticated user -- see next bullet (# 2-ii) for exception.
   2. The "you missed this" popup for the participation release checkbox is in the totally wrong place for me, even as non-admin (I double-checked -- there was no other checkbox off-screen for it to be trying to point to)-- Firefox:
   ![image](https://cloud.githubusercontent.com/assets/2660348/25241682/e1595b3c-25c4-11e7-8b38-7c995d0e6c8c.png)
      1. NOTE: "please check this box" popup for Participation Release checkbox field is NOT showing up in Jess's tests (Chrome and IE)!  From Jess (excerpt; [see her comment](#issuecomment-295846397) > 4th checkbox for more info):
      > There is no message to the user that they must check the box. This is on both Mac and Win 7 using Chrome while logged in as an authenticated user.
   3. It's crummy that it only points to one missed field -- BUT, would be a little less crummy if the style-treatment of missed required fields were better (i.e. if select list fields weren't the only ones with red halos), and if the "required asterisks" were more conspicuous.
      1. As mentioned as an idea in the "task" checkboxes, maybe adding an overall "Please fill in all required fields" message at the top/bottom of the form would be helpful -- and maybe important for accessibility...?
3. But most importantly, required checkbox fields AND text fields do not get highlighted as being missed required fields if you submit the form without filling them in.
![image](https://cloud.githubusercontent.com/assets/2660348/25242395/6c3bfe92-25c7-11e7-9a6b-e620e0f96305.png)

### Background
_(from Jess)_

I went through the form and filled out all fields marked as required. Clicking on Register button did nothing.  I assumed it was because i missed a required field, but there was no indicator on the form of what was missing.

I started going through field by field to try to submit. I was able to submit once I selected something under Participation Release.

The form needs better indicators of required fields as well as notifications of what was required that wasn't completed when a user clicks submit (ideally taking them top top of page with a message about required fields missing). 

Right now you click the button and nothing happens.

### ALSO...
Background from #294 -- The "attendee email field" is marked as required yet the descriptive text allows for you to leave it blank.
- Field should not be required (if empty, email goes to logged in user's email address).
- Descriptive text should be clearer:
_Enter the e-mail address of the recipient who will use this ticket. Leave blank to receive a redemption code for this registration that you can then forward on to whomever you like._
- FYI Jess also updated #213 with new text.

**Comment by aweingarten** [comment]
Note please also address anything mentioned here:
https://github.com/Drupal4Gov/Drupal-GovCon-2017/issues/294

**Comment by lkopacz** [comment]
@jdearie I am unable to recreate this error, but we are not supposed to be seeing the standard Drupal Error that was used in Drupal 7. The webform module uses HTML5 which has different default error handling than Drupal's former Handling. It goes through every item that you missed, one by one, if you missed it. 

CC @aweingarten @alisonjo2786 

<img width="1293" alt="screen shot 2017-04-17 at 9 48 26 pm" src="https://cloud.githubusercontent.com/assets/6540866/25111249/c9c39e6a-23b7-11e7-9da3-8a3a1e6be26a.png">

**Comment by alisonjo2786** [comment]
@aweingarten @lkopacz Updated issue description with additional details and list of reqs as I see them -- @jdearie pls feel free to re-edit, of course!

**Comment by alisonjo2786** [comment]
@aweingarten High priority

**Comment by alisonjo2786** [comment]
More re: "post-submit"/"you missed these required fields" -- 

Here's the treatment of a select list:
![image](https://cloud.githubusercontent.com/assets/2660348/25196843/e3b9f0e0-250f-11e7-8f3a-52aba413af05.png)

But text box and checkbox fields are not given any red highlighting or whatever.

**Comment by alisonjo2786** [comment]
@jdearie I moved #294 over to this ticket, per request from @aweingarten.
- I added a vague task/checkbox to the top of the "requirements" list.
- I copied the issue details into a section at the bottom of the issue description, "ALSO..."

**Jess action:** Please review the checkbox I added to the top of the requirements list, and please add a screenshot of this field and the problematic label/description/required-ness under the "ALSO..." heading at the bottom.

**Comment by jdearie** [comment]
I took another look and tried registering for another ticket... the required fields notification works for nearly all the required fields.

I answered every field marked with an asterisk... the only thing I didn't do was complete non-required fields and the Participation Release section.  Clicking the Register button did nothing.  I didn't get a notification about required fields like I did for the required fields (and as shown in the screen shot provided by @lkopacz). 

HOWEVER - once I checked the checkbox saying I would adhere to the CoC and Media Policy, I was able to register.

Since that checkbox is required - I think we need to either break it out somehow or figure out why users aren't getting any kind of message telling them that one field is required.  Perhaps remove it from the Participation Release field collection and list it separately?

@alisonjo2786 - the checkbox items look good to me and appear to be mostly fixed now.
not sure what you mean by screen shot the ALSO ... my only comment was that the Attendee email address field IS NOT required and that the descriptive/helper text should read "Enter the e-mail address of the recipient who will use this ticket. Leave blank to receive a redemption code for this registration that you can then forward on to whomever you like."  - thus making it more clear what that field is for.

**Comment by alisonjo2786** [comment]
[tagged up with Jess outside github, she's going to post another update and screenshot in a bit]

**Comment by jdearie** [comment]
ok so. a few updates:

- [ ] Attendee email descriptive/helper text should read "Enter the e-mail address of the recipient who will use this ticket. If you are registering for someone else, please enter their email address."  Keep it a required field.

- [ ] Food preferences field contains both "none" and "no restrictions".  Let's remove None, and replace the "-None-" default with "-Select-" that we have "-Select-" and "No Restrictions" before the other options.

- [ ] Missing word in the Which days do you plan field collection. Text should read:
This helps us get an accurate count for lunch. Don't worry, you can still attend all three days regardless of what you select here. Give us your best guess.
_ current the word **you** before the word **select** is missing_

- [ ] While the Participation Release field collection contains ONE required field, and that field is correctly marked with an asterisk, it does not behave like other required fields when it is not checked. There is no message to the user that they must check the box.  This is on both Mac and Win 7 using Chrome while logged in as an authenticated user.  There's nothing to screen shot since nothing is happening... lol - that's the issue.  I would recommend we remove the required CoC/Media Policy field from the Participation Release field collection and make it a stand-alone required field that, when left blank, shows the user a message telling them they have to fill out the required field.  

I updated #213 to reflect these items.

**Comment by alisonjo2786** [comment]
_re: Comment from @jdearie (directly above this):_

I'm checking off boxes when I've confirmed that the items are covered in this ticket ( #292 ) or some other ticket.
- First checkbox: Updated this issue ( #292 ) description to reflect specs in Jess's comment.
- Second: Updated #314 with specs in Jess's comment here.
- Third: Added new GH issue: #316
- Fourth: Updated this issue ( #292 ) description to reflect specs in Jess's comment.
